src/Storages/MergeTree/PartMetadataManagerWithCache.cpp

```diff
@@ -0,0 +1,273 @@
+#include "PartMetadataManagerWithCache.h"
+
+#if USE_ROCKSDB
+#include <Common/hex.h>
+#include <Common/ErrorCodes.h>
+#include <IO/HashingReadBuffer.h>
+#include <IO/ReadBufferFromString.h>
+#include <Storages/MergeTree/IMergeTreeDataPart.h>
+
+namespace ProfileEvents
+{
+    extern const Event MergeTreeMetadataCacheHit;
+    extern const Event MergeTreeMetadataCacheMiss;
+}
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int CORRUPTED_DATA;
+    extern const int NO_SUCH_PROJECTION_IN_TABLE;
+}
+
+PartMetadataManagerWithCache::PartMetadataManagerWithCache(const IMergeTreeDataPart * part_, const MergeTreeMetadataCachePtr & cache_)
+    : IPartMetadataManager(part_), cache(cache_)
+{
+}
+
+String PartMetadataManagerWithCache::getKeyFromFilePath(const String & file_path) const
+{
+    return disk->getName() + ":" + file_path;
+}
+
+String PartMetadataManagerWithCache::getFilePathFromKey(const String & key) const
+{
+    return key.substr(disk->getName().size() + 1);
+}
+
+std::unique_ptr<SeekableReadBuffer> PartMetadataManagerWithCache::read(const String & file_name) const
+{
+    String file_path = fs::path(part->getFullRelativePath()) / file_name;
+    String key = getKeyFromFilePath(file_path);
+    String value;
+    auto status = cache->get(key, value);
+    if (!status.ok())
+    {
+        ProfileEvents::increment(ProfileEvents::MergeTreeMetadataCacheMiss);
+        auto in = disk->readFile(file_path);
+        readStringUntilEOF(value, *in);
+        cache->put(key, value);
+    }
+    else
+    {
+        ProfileEvents::increment(ProfileEvents::MergeTreeMetadataCacheHit);
+    }
+    return std::make_unique<ReadBufferFromOwnString>(value);
+}
+
+bool PartMetadataManagerWithCache::exists(const String & file_name) const
+{
+    String file_path = fs::path(part->getFullRelativePath()) / file_name;
+    String key = getKeyFromFilePath(file_path);
+    String value;
+    auto status = cache->get(key, value);
+    if (status.ok())
+    {
+        ProfileEvents::increment(ProfileEvents::MergeTreeMetadataCacheHit);
+        return true;
+    }
+    else
+    {
+        ProfileEvents::increment(ProfileEvents::MergeTreeMetadataCacheMiss);
+        return disk->exists(fs::path(part->getFullRelativePath()) / file_name);
+    }
+}
+
+void PartMetadataManagerWithCache::deleteAll(bool include_projection)
+{
+    Strings file_names;
+    part->appendFilesOfColumnsChecksumsIndexes(file_names, include_projection);
```

Comment:
What do you think about using getByPrefix instead? This would simplify this a lot.

Reply:
>and it is more efficient in performance because appendFilesOfColumnsChecksumsIndexes get all metadata files without visiting disk

Yes, I also thought about this, but code to obtain all files quite complex and I guess that rocksdb db will be in cache always since it will be pretty small, so the performance different unlikely will be significant.

But it is up to you, to decide of course.

By the way, maybe we can get all files from checksums.txt?

Reply:
I am not sure file names in `checksums.txt` is complete, especially before parts are commited. 
And files in `checksums.txt` include not only metadata, but also other data file, if we want to filter out metadata files from `checksums.txt`, we must know list of metadata files firstly...

Reply:
>I am not sure file names in checksums.txt is complete, especially before parts are commited.

AFAICS the call chain is as follow `IMergeTreeDataPart::renameTo` -> `PartMetadataManagerWithCache::updateAll` -> `appendFilesOfColumnsChecksumsIndexes`, and in `renameTo`, `checksums.txt` should be written completely.

>And files in checksums.txt include not only metadata, but also other data file, if we want to filter out metadata files from checksums.txt, we must know list of metadata files firstly...

Ok, agree, thanks for clarification.

Reply:
@azat  I removed some unless code from PR in lastest commit. 
As for the choice whether using `appendFilesOfColumnsChecksumsIndexes` or `getByPrefix`, even if I change to `getByPrefix` in `PartMetadataManagerWithCache::deleteAll`,  I still have to use `appendFilesOfColumnsChecksumsIndexes` in `PartMetadataManagerWithCache::updateAll`, because metadata is not always in cache. See this code in `IMergeTreeDataPart::renameTo`:  
```
    metadata_manager->deleteAll(true);
    metadata_manager->assertAllDeleted(true);
    volume->getDisk()->setLastModified(from, Poco::Timestamp::fromEpochTime(time(nullptr)));
    volume->getDisk()->moveDirectory(from, to);
    relative_path = new_relative_path;
    metadata_manager->updateAll(true);
```
To keep code style consistency of PartMetadataManagerWithCache, I think maybe keep code in `deleteAll` is better, what do you think ?

Reply:
>To keep code style consistency of PartMetadataManagerWithCache, I think maybe keep code in deleteAll is better, what do you think ? 

Completely agree - the code should be the same, thanks!